### ISSUE TITLE
[RFC] Unified hook API

### ISSUE THREAD

**Issue by jwx** [aurelia/aurelia#1047]
# 💬 RFC

This RFC proposes a unified API to create hooks that can be applied for components and framework events. Initial uses are within component lifecycle and routing.

This API will be a direct replacement for v1's `@viewEngineHooks`, as well as v1 router's pipeline slots, with improved consistency and leaving room for (potentially unknown) future enhancements.

Unlike `@viewEngineHooks`, there is not a pre-determined set of hooks that can be tapped into. This allows users and plugin authors to define new hooks if they want to, in much the same way that Aurelia itself would be utilizing it.

**This proposal is a merge of the proposals in https://github.com/aurelia/aurelia/issues/307#issuecomment-445824916 (and related) and https://github.com/aurelia/aurelia/issues/1044.**

### Might address these issues and/or is related to
- https://github.com/aurelia/aurelia/issues/978
- https://github.com/aurelia/aurelia/issues/1037

## TL;DR (what does the API look like?)

A hook is created either by
- convention, a class has the suffix `Hook` (only applies to classes),
- a decorator, `@hook`, or
- a function call, `Hook.define()`.

```js
export class DoStuffHook {
  binding(vm, ...args) {}
  attaching(vm, ...args) {}
  detaching(vm, ...args) {}
  // etc
}

@hook()
export class DoStuff {
  binding(vm, ...args) {}
  attaching(vm, ...args) {}
  detaching(vm, ...args) {}
  // etc
}

const animateIn = Hook.define('attaching', (vm, ...args) => {});

const auth = Hook.define({ name: 'canLoad', exclude: ['login-page', 'help-page'] }, (vm, ...args) => {});
```
Once a hook is created it should be `register`ed to apply within the context of the registration.
```js
export class ConsumerCustomElement {
  static dependencies = [DoStuff]; // Local registration
}

Aurelia
  .register( // Global registration
    animateIn, // From above example
    Hook.define({ name: 'canLoad', exclude: ['login-page', 'help-page'] }, (vm, ...args) => {}),
  )
  .app(MyApp).start();
```

### Use cases
A common requirement in apps is to run the same bit of logic in the same hook in multiple components. Some examples:
1) `*`: Conditionally log each hook invocation for advanced troubleshooting.
2) `binding`, `load`: Lazy load data from the server only once, but initiate this from whichever component (out of several that need it) happens to be rendered first.
3) `define`, `hydrating`, `hydrated`, `created`: Apply CSS to components matching arbitrary/dynamic criteria or other custom runtime conventions
4) `attaching`, `detaching`: Run the same enter/leave animations for all routed components.
5) `canLoad`: Run auth guards for all but a specific set of excluded components.
6) `canUnload`: Run an "unsaved changes" dialog guard for all components that have an `isDirty` getter on them.
7) `transformTitle`: Translate all component titles except 'profile' (with user's name as title) before updating the browser's title.

### Resource semantics

When `@hook` is applied to a class or a function, that class or function becomes a "resource" (it's registered to the `Protocol.resource` metadata api just like `@customElement`, `@valueConverter`, etc), giving it resource-specific semantics for DI and the runtime module loader/analyzer:
- Registering it in the root DI container (e.g. via `Aurelia.register`) turns it into a global resource and causes it to be invoked for *every* component
- Adding it to the `dependencies` list of a specific component, causes it to be invoked for only that component.
- It's recognized by the module analyzer, meaning it can lazy loaded (natively or with help of webpack / other bundlers) with relative ease.
- It will be supported by conventions (by adding the `Hook` suffix to the class name).
```ts
// explicit
@hook()
export class SharedStuff { ... }

// conventions:
export class SharedStuffHook { ... }
```

### Configuration

A hook can be configured when created to, for example, include or exclude components from the hook. The options that can be provide to the `@hook` decorator or the `Hook.define` call are:
- name - specifies which hook a function should be applied to. Mandatory for function hooks, not allowed for class hooks.
- include - specifies which components, or components loaded in a viewport, the hook applies for.
- exclude - specifies which components, or components loaded in a viewport, the hook does not apply to.

The `include/exclude` options can be either
- a component name (string)
- a component type (CustomElement object)
- an object with at least one of
  - `component:` a component name or type
  - `viewport:` a viewport name or a `Viewport` instance
- a boolean function deciding inclusion that is called for each possibly affected object
- a list mixed with any of the above (array)

The `include` option is applied before the `exclude` option.

```js
// This hook applies to everything in the registration context (global or local)
Hook.define('detaching', (vm, ...args) => { /* play Pac-Man death sound */ });

// The hooks in the class only applies to the four ghosts regardless of registration context
@hook({ include: ['Inky', Blinky, { component: 'Pinky' }, { component: Clyde }] })
class TrueBlue { /* Run away! */
  attaching(vm, ...args) {}
  detaching(vm, ...args) {}
}

// The hook applies to everything in the viewport except the excluded components regardless of registration context
Hook.define({
  name: 'canLoad',
  include: { viewport: 'secured' },
  exclude: ['no-access', ContactSupport ]},
  (vm, ...args) => { /* Check auth */ });
```

### Contracts & behavior

- The first argument passed-in is always the related object, for example the view model for a lifecycle hook, followed by the arguments that are normally passed-in to the hook.
- `Hook`s are *scoped singletons*, just like `ValueConverter`s and `BindingBehavior`s, meaning globally registered ones are global singletons, whereas locally registered ones are singletons only within their registered context (typically one instance per type)
- Shared hooks run *before* (and if async, in parallel with) their component instance and framework counterparts.

### Future extensions

Depending on the feedback during alpha, we may add some quality-of-life configuration options to the `@hook` to enable more use cases and/or improve performance. For example:
```ts
// Control whether the hook runs before, after, or in parallel with the component instance hook.
// This would default to 'parallel' (which is the current only behavior), making this a non-breaking change.
@hook({ order: 'before' | 'after' | 'parallel' })
```

## 💻 Examples

Let's go through the aforementioned use cases, now with example implementations using this new API:
### 1) `*`

Conditionally log each lifecycle hook invocation for advanced troubleshooting.

- Registration scope: global

```js
// lifecycle-logger.js
@hook()
export class LifecycleLogger {
  constructor(@ILogger logger) { this.logger = logger; }
  define(vm) { this.trace('define', vm); }
  hydrating(vm) { this.trace('hydrating', vm); }
  hydrated(vm) { this.trace('hydrated', vm); }
  created(vm) { this.trace('created', vm); }
  binding(vm) { this.trace('binding', vm); }
  bound(vm) { this.trace('bound', vm); }
  attaching(vm) { this.trace('attaching', vm); }
  attached(vm) { this.trace('attached', vm); }
  detaching(vm) { this.trace('detaching', vm); }
  unbinding(vm) { this.trace('unbinding', vm); }
  canLoad(vm) { this.trace('canLoad', vm); return true; }
  load(vm) { this.trace('load', vm); }
  canUnload(vm) { this.trace('canUnload', vm); return true; }
  unload(vm) { this.trace('unload', vm); }
  trace(hook, vm) { this.logger.trace(`${hook} ${vm.$controller.definition.name}`); }
}

// index.js
Aurelia
  .register(
    LifecycleLogger,
    ...,
  )
  .app(...).start();
```

### 2) `binding`, `load`

Lazy load data from the server only once, but initiate this from whichever component (out of several that need it) happens to be rendered first.

- Registration scope: local (added by hand to `dependencies`)

(in this example, we use it for a dropdown that needs some server-side data that only needs to be loaded once)

```js
// things-initializer.js
export const ICommonThings = DI.createInterface().withDefault(x => x.singleton(CommonThings));
export class CommonThings {
  constructor(@IHttpClient http) { this.http = http; }
  init() {
    if (!this.isInitialized) {
      return this.initPromise ??= (async () => {
        this.data = await this.http.get('api/common-things');
        this.isInitialized = true;
      })();
    }
  }
}
@hook
export class ThingsInitializer {
  constructor(@ICommonThings things) { this.things = things; }
  binding() { return this.things.init(); }
  load() { return this.things.init(); }
}
// some-dropdown.js
@customElement({
  name: 'some-dropdown',
  template: '<select :value="thingId"><option repeat.for="thing of things.data" :model="thing.id">${thing.name}</option></select>',
  dependencies: [ThingsInitializer],
})
export class SomeDropdown {
  constructor(@ICommonThings things) { this.things = things; }
}
```

### 3) `define`, `hydrating`, `hydrated`, `created`

Apply CSS to components matching arbitrary/dynamic criteria or other custom runtime conventions

- Registration scope: global

```js
// ripple-effect-applicator.js
@hook
export class RippleEffectApplicator {
  created(vm) {
    if (vm.$controller.definition.name.startsWith('bs-') && vm.someCondition) {
      // (note: this definition.instructions stuff will become easier to use soon)
      vm.$controller.definition.instructions[0].push(new HydrateAttributeInstruction('ripple', []));
    }
  }
}

// index.js
Aurelia
  .register(
    RippleEffectApplicator,
    ...,
  )
  .app(...).start();
```

### 4) `attaching`, `detaching`:

Run the same enter/leave animations for all components.

- Registration scope: local (added by custom decorator)

```js
// animations.js
const duration = 300;
@hook
export class SlideInOutAnimations {
  attaching(vm) {
    return vm.$controller.host.animate([{ transform: 'translate(-100vw, 0)' }, { transform: 'translate(0, 0)' }], { duration, easing: 'ease-in' }).finished;
  }
  detaching(vm) {
    return vm.$controller.host.animate([{ transform: 'translate(0, 0)' }, { transform: 'translate(100vw, 0)' }], { duration, easing: 'ease-out' }).finished;
  }
}
@hook
export class FadeInOutAnimations {
  attaching(vm) {
    return vm.$controller.host.animate([{ opacity: 0 }, { opacity: 1 }], { duration }).finished;
  }
  detaching(vm) {
    return vm.$controller.host.animate([{ opacity: 1 }, { opacity: 0 }], { duration }).finished;
  }
}
// let's make a convenient decorator for ourselves this time:
export function pageTransitions(target) {
  (target.dependencies ??= []).push(SlideInOutAnimations, FadeInOutAnimations);
}

// home-page.js
@pageTransitions
export class HomePage {}

// settings-page.js
@pageTransitions
export class SettingsPage {}
```

### 5) `canLoad`:

Run auth guards for all but a specific set of excluded components.

- Registration scope: global

```js
// auth-guard.js
@hook({ exclude: ['login', 'recover-password'] })
export class AuthGuard {
  constructor(@IAuthService authService) { this.authService = authService; }
  // (note: router hook signatures are still subject to change)
  async canLoad(vm, params, instruction) {
    if (await this.authService.isAuthorized(instruction.data.auth)) {
      return true;
    }
    return `/login(reason=${instruction.data.auth.scopes})`;
  }
}

// presentation.js
// Check standard authorization
export class PresentationPage {}

// settings-page.js
// Check standard authorization
export class SettingsPage {}

// profile-page.js
// Check 'profile' authorization
@route({ data: { auth: ['profile'] } })
export class ProfilePage {}

// index.js
Aurelia
  .register(
    AuthGuard,
    ...,
  )
  .app(...).start();
```

### 6) `canUnload`

Run an "unsaved changes" dialog guard for all components that have an `isDirty` getter on them.

- Registration scope: global

```js
// dirty-state-guard.js
@hook
export class DirtyStateGuard {
  constructor(@IDialogService dialog) { this.dialog = dialog; }
  async canUnload(vm,  params, instruction) {
    if (vm.isDirty) {
      return this.dialog.promptUnsavedChanges();
    }
    return true;
  }
}

// index.js
Aurelia
  .register(
    DirtyStateGuard,
    ...,
  )
  .app(...).start();
```

### 7) `transformTitle`

Translate all component titles except 'profile' (with user's name as title) before updating the browser's title.

- Registration scope: global

```js
// translate-title.js
const TranslateTitle = Hook.define({
    name: 'transformTitle',
    exclude: 'profile',
  },
  title => typeof title === 'string' ? translate(title) : title
);

// books.js
@title('My books')
class Books {}

// store.js
class Store {} // Automatically gets title 'Store'

// profile.js
@title(vm => vm.profile.fullName)
class Profile {}

// index.js
Aurelia
  .register(
    TranslateTitle,
    ...,
  )
  .app(...).start();
```
